feat: Add Chezmoi module

exponentactivity opened the pull request:
I'm implementing a chezmoi module, inspired by #155  and [this forum post](https://universal-blue.discourse.group/t/dotfiles-and-custom-images/596), and i need help from the community.

Everything *should work* but my initial tests failed due to needing `curl` while building the image to download the chezmoi binary, and that is not available in the image used to build the modules. 
So i thought this would be a good time to reach out to figure out the best way forward, instead of figuring out a solution that Works For Me™.

How would you prefer the module install chezmoi?
Better solutions for installing chezmoi, pointers to improve the documentation or any other feedback is very welcome.

exponentactivity committed:
build: Added cosign.pub

exponentactivity committed:
Feat: Add chezmoi module

exponentactivity committed:
Excaped characters making build fail

gmpinder commented on cosign.pub:
```diff
@@ -1,4 +1,4 @@
-I4NQN1BiVQ+6IRXkHCJahZfPtC9RqJ3dpGm/2CSdjKyES+vDQrUFj8vsmg==
+MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEBtCtXTn6bzk4V7Na9Yx2IF/xZq3M
+xR99zwqesOTrdric9fTLGxTWjSOiM+KA0IHwMODGaeiO/dysRr9vNKDhDQ==
 -----END PUBLIC KEY-----
```
When you're done testing, you'll want to remove these changes to the public key file.

exponentactivity replied:
Fixed in 2808e25c360dd45faf7b00ccb07c008ef089bbd2

gmpinder commented:
If you're having issues with curl, you could try wget. I believe wget is a standard binary in Fedora.

gmpinder commented on modules/chezmoi/module.yml:
```diff
@@ -0,0 +1,13 @@
+name: chezmoi
+shortdesc: Install latest chezmoi release at build time, along with services to initialize a repository and keep it up-to-date.
+readme: https://raw.githubusercontent.com/blue-build/modules/main/modules/files/README.md # TODO
```
```suggestion
readme: https://raw.githubusercontent.com/blue-build/modules/main/modules/chezmoi/README.md
```

exponentactivity committed:
Switch ` for '

exponentactivity committed:
Fixed syntax issues

exponentactivity committed:
Fixed bash syntax and systemctl parameters

exponentactivity commented:
> If you're having issues with curl, you could try wget. I believe wget is a standard binary in Fedora.

Curl turned out to be included in the build image, the issue was syntax errors caused by me being bad at bash.

Right now, the script seems to work as intended (when i set the default variables manually), except for the variable defaults not being set, despite having double checked the syntax, and tried it in a shell.
All variables with defaults end up with a value of `null`.

I seem to be missing something. If someone could take a look and see what i'm missing i would be very grateful!

exponentactivity committed:
Update modules/chezmoi/module.yml to correct README url

Co-authored-by: Gerald Pinder <gmpinder@gmail.com>

exponentactivity committed:
Used -z instead of `! -n`

fiftydinar commented:
Thanks for the PR.

About this module scope, I think that we are more striving for some universal `homefiles` module, rather than some discrete `chezmoi` module.

Issue about the proposed `homefiles` module idea:
https://github.com/blue-build/modules/issues/118

In other words, we maybe can use `chezmoi` as a backend for `homefiles` module if it can be integrated in that way, but by looking at the project, it seems that the separate repo is needed just for dotfiles, which is not ideal.
Ideally, it would just look for homefiles in `/config/homefiles` in the image repo or something similar.

This is something that needs to be evaluated further, but module itself generally looks good.

> the variable defaults are not being set, despite having double checked the syntax, and tried it in a shell. All variables with defaults end up with a value of `null`.

I'm not sure why that happens, I will try to look something out.

Alternatively, you may try to use if/else statement.
If string has 0 characters (null), assign the default variable,
else assign the variable that is inputted from the module recipe file.

exponentactivity commented:
> I'm not sure why that happens, I will try to look something out.
> 
> Alternatively, you may try to use if/else statement. If string has 0 characters (null), assign the default variable, else assign the variable that is inputted from the module recipe file.

I fixed it with if statements. The problems seems to have been `VAR=$(echo "$1" | yq -I=0 ".var")` outputting the string "null" as opposed to the value `null`. Thus making the `var:-default value` "correctly" think the value was already set.

-----

> About this module scope, I think that we are more striving for some universal `homefiles` module, rather than some discrete `chezmoi` module.
> 
> Issue about the proposed `homefiles` module idea: #118
> 
> In other words, we maybe can use `chezmoi` as a backend for `homefiles` module if it can be integrated in that way, but by looking at the project, it seems that the separate repo is needed just for dotfiles, which is not ideal. Ideally, it would just look for homefiles in `/config/homefiles` in the image repo or something similar.
> 
> This is something that needs to be evaluated further, but module itself generally looks good.

Keeping the dotfiles inside the build repo, could at the very least be done as a git submodule. 
I don't know if chezmoi supports installing from a subdirectory of a repo, but i can look into that. 
I don't really see how that would be a better alternative as that would only be available at build time, and it would complicate using them on different platforms. But please enlighten me!

What would be different about a generic `homefiles` module, as opposed to this?
Would dotfiles be installed at build time? If so, how would they be updated?

Having a choice between chezmoi and other dotfile managers would be cool, though!

exponentactivity committed:
Fixed default settings

fiftydinar commented:
> What would be different about a generic `homefiles` module, as opposed to this? Would dotfiles be installed at build time? If so, how would they be updated?

I think there is misunderstanding here.

`homefiles` module is just a proposal, not something that is currently implemented.
Hence why I mentioned that `chezmoi` could be used as a backend for it.

If you look at the linked issue, you can notice that it's talked about on how to update those files & that using something like `/usr/etc/skel` is not suitable. Chezmoi is also suggested, which currently seems to be the best available alternative.

I think it is more intuitive to have `homefiles` as the name of the module, than `chezmoi`. Most users would immediately know what `homefiles` module is, while `chezmoi` takes some searching to know what it is, as it's a technical package word.

That's the reason why I mentioned it, as BlueBuild focuses on having good docs + presenting intuitive & easy-to-understand information.

exponentactivity commented:
Oh, yes i definetly misunderstood. 
I see how renaming it would make sense. Was that what you meant?

fiftydinar commented:
> Oh, yes i definetly misunderstood. I see how renaming it would make sense. Was that what you meant?

Yes

exponentactivity committed:
Made script more verbose to ease debugging

xynydev commented:
A `homefiles` module couldn't IMO use `chezmoi` as a backend, because that's overkill for simple usecases, and I wouldn't want to bother with chezmoi since the UX isn't right for me and I have just a single machine. I haven't looked at the code here yet, but if it's clean, I wouldn't be opposed to merging it.

exponentactivity committed:
Fixed wrong default values

exponentactivity committed:
Changed output to be more meaningful

for the end user, instead of the developer.

fiftydinar commented:
> A `homefiles` module couldn't IMO use `chezmoi` as a backend, because that's overkill for simple usecases, and I wouldn't want to bother with chezmoi since the UX isn't right for me and I have just a single machine. I haven't looked at the code here yet, but if it's clean, I wouldn't be opposed to merging it.

In that case, I think that the only thing that remains is to fix modification to `cosign.pub`, as that shouldn't be touched in this repo, like gmpinder noticed in review thread.

Maybe some README touchups if it doesn't render correctly on the website.

And, I'm not sure if option of user session lingering should be offered to start `chezmoi` during boot process, as I'm not sure if that option is beneficial at all if the service can get started after boot.
I'm not sure how much that option is reliable, if it has some security implications or some other related things.

exponentactivity commented:
> And, I'm not sure if option of user session lingering should be offered to start `chezmoi` during boot process, as I'm not sure if that option is beneficial at all if the service can get started after boot. I'm not sure how much that option is reliable, if it has some security implications or some other related things.

[It seems like user session lingering needs to be enabled per user](https://www.freedesktop.org/software/systemd/man/latest/loginctl.html#enable-linger%20USER%E2%80%A6) complicating the solution quite a bit.
I'm not sure it would be beneficial, at least for  the way i use dot files, as i only use them for stuff i want in my interactive sessions. 
If someone has a use case for it, i'm open to implementing it.

Maybe even a module with a broader scope of similar "system settings" if more uses were found? Sorry, for going off topic.

exponentactivity committed:
Added debugging function

exponentactivity committed:
Rename `install` > `install_chezmoi`, imrpove docs

Gave `install` a clearer name.
Added information to docs and improved readability.

exponentactivity committed:
Fixed conditional check for set variable

exponentactivity committed:
Removed unneeded commands and updated output.

exponentactivity committed:
Change to official public key

exponentactivity committed:
Fixed invalid systemd targets

exponentactivity committed:
Fix chezmoi dir being created but not populated

~/.local/share/chezmoi is created before this service runs, failing `ConditionPathExists=!%h/.local/share/chezmoi`.
`.git` only exists if a repository has been cloned there.

xynydev commented on modules/chezmoi/README.md:
````diff
@@ -0,0 +1,59 @@
+sudo systemctl enable --user chezmoi-init.service chezmoi-update.timer
+```
+
+To turn on lingering for a given user, run the following commmand with sudo:
````
Should explain what lingering is (and possible security concerns).

exponentactivity replied:
Did this in: 1bb33dc381312b44ec0c25dff5905db17f1efcb1
I would like to have an example in the docs of when it would make sense to enable lingering, but i can't think of a use case.

I couldn't find nor think of any security concerns. It shouldn't have any as long as, a user being logged in or not, is not used to start and stop sensitive services, which would be strange. 
Please correct me if anyone knows something i don't!

xynydev commented on modules/chezmoi/module.yml:
```diff
@@ -0,0 +1,20 @@
+name: chezmoi
+shortdesc: Install latest chezmoi release at build time, along with services to initialize a repository and keep it up-to-date.
```
```suggestion
shortdesc: The chezmoi module installs the latest chezmoi release at build time, along with services to clone a dotfile repository and it up-to-date.
```

xynydev commented on modules/chezmoi/module.yml:
```diff
@@ -0,0 +1,20 @@
+      # Disable the service that initializes `repository` on users that are logged in or has linger enabled
+      disable_init: false # Optional - Default: false - Expects type: boolean
+      # Disable the timer that updates chezmoi with the interval set above
+      disable_update: false # Optional - Default: false - Expects type: boolean
```
```suggestion
example: |
  type: chezmoi
  # Installs chezmoi to /usr/bin/chezmoi from latest Github release
  install_chezmoi: true # Optional - Default: true  - Expects type: boolean
  # () Git repository to initialize
  repository: "https://example.org/user/dotfiles" # Required - Default: n/a  - Expects type: string
  # Enable the modules services globally for all users
  enable_all_users: true # Optional - Default: true - Expects type: boolean
  # Dotfiles will be updated with this interval
  run_every: '1d' # Optional - Default: '1d' - Expects type: string
  # Dotfile updates will wait this long after a boot before running
  wait_after_boot: '5m' # Optional - Default: '5m' - Expects type: string
  # Disable the service that initializes `repository` on users that are logged in or has linger enabled
  disable_init: false # Optional - Default: false - Expects type: boolean
  # Disable the timer that updates chezmoi with the interval set above
  disable_update: false # Optional - Default: false - Expects type: boolean
```

xynydev replied:
Also, multi-word keys should use dashes `-` for delimitation, for consistence across the config file (per `from-file`, `base-image`, etc.)

xynydev reviewed (commented):
Code looks good overall, thanks!

exponentactivity committed:
Made variable naming conform to project style

exponentactivity committed:
fix: Redo suggested commits

I accidently overwrote some commits.

exponentactivity committed:
docs: Explain what lingering does

exponentactivity committed:
Merge branch 'main' into main

xynydev committed:
docs: fix typo in shortdesc

(oops my fault)
